clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -111,12 +126,401 @@ class image_mem_wrapper {
+    if (_res_D3D11_set.find(_res_D3D11) != _res_D3D11_set.end()) {
+      throw std::runtime_error("This D3D11 resource is already registered!");
+    }
+
+    // Query the information from D3D11 resource object
+    query_res_info(_res_D3D11);
+
+    // Create a shared handle for the DX11 resource to enable interop.
+    _res_win_nt_handle = create_shared_handle(_res_D3D11);
+
+    // Prepare the external memory descriptor for SYCL interop.
+    sycl::ext::oneapi::experimental::resource_win32_handle ext_mem_win_handle{
+        _res_win_nt_handle};
+
+    sycl::ext::oneapi::experimental::external_mem_descriptor<
+        sycl::ext::oneapi::experimental::resource_win32_handle>
+        ext_mem_desc{ext_mem_win_handle,
+                     sycl::ext::oneapi::experimental::external_mem_handle_type::
+                         win32_nt_dx12_resource,
+                     _res_size_bytes};
+
+    // Import the external memory into SYCL for use with SYCL APIs.
+    new (_res_interop_mem_handle_ptr)
+        sycl::ext::oneapi::experimental::interop_mem_handle(
+            sycl::ext::oneapi::experimental::import_external_memory(
+                ext_mem_desc, get_default_queue()));
+
+    // Register the resource to track its uniqueness.
+    _res_D3D11_set.insert(_res_D3D11);
+  }
+
+  /// Cleans up the interop wrapper by releasing the SYCL interop
+  /// memory handle and the shared handle created for the DX11 resource.
+  virtual ~interop_mem_wrapper() {
+    if (!_res_interop_mem_handle_ptr) {
+      throw std::runtime_error(
+          "Resource is not registered! "
+          "Resource should be registered before unregistration.");
+    }
+
+    // Close the shared handle to the DX11 resource.
+    CloseHandle(_res_win_nt_handle);
+
+    // Release the external memory associated with the SYCL interop handle.
+    sycl::ext::oneapi::experimental::release_external_memory(
+        get_interop_mem_handle(), get_default_queue());
+
+    // Remove the resource from the tracking set.
+    _res_D3D11_set.erase(_res_D3D11);
+  }
+
+private:
+  static std::set<ID3D11Resource *> _res_D3D11_set;
+
+  ID3D11Resource *_res_D3D11 = nullptr;
+  HANDLE _res_win_nt_handle = nullptr;
+  unsigned _res_reg_flags = 0;
+  sycl::ext::oneapi::experimental::interop_mem_handle
+      *_res_interop_mem_handle_ptr = nullptr;
+
+  /// Helper function to query the properties of DX11 resource
+  void query_res_info(ID3D11Resource *resource) {
+    unsigned int res_width = 1;
+    unsigned int res_height = 1;
+    unsigned int res_depth = 1;
+    unsigned int res_num_levels = 1;
+    unsigned int res_arr_size = 1;
+
+    sycl::image_channel_type img_ch_type =
+        sycl::image_channel_type::unsigned_int8;
+
+    image_channel channel;
+
+    // Get the dimension info of DX11 resource
+    D3D11_RESOURCE_DIMENSION dimension;
+    resource->GetType(&dimension);
+
+    switch (dimension) {
+    case D3D11_RESOURCE_DIMENSION_BUFFER: {
+      _res_is_buffer = true;
```

Comment:
Cannot access the private member in the base class.

Reply:
Made it a protected member.